source/game/field/obj/ObjectHeyho.cc

```diff
@@ -0,0 +1,247 @@
+    : ObjectCollidable(params), StateManager(this), m_color(params.setting(1)) {
+    const auto *rail = RailManager::Instance()->rail(params.pathId());
+    ASSERT(rail);
+    const auto &railPts = rail->points();
+
+    // The two endpoints are candidates for the highest Y coordinate in the route
+    // The midpoint's Y coordinate should be the lowest in the route
+    f32 startY = railPts.front().pos.y;
+    f32 endY = railPts.back().pos.y;
+    m_apex = startY >= endY ? startY : endY;
+    m_midpoint = railPts[railPts.size() / 2].pos;
+
+    // The object should speed up as we approach the low point and slow down as we leave it
+    // We form an acceleration constant so multiplying with (pos - center) gives v^2
+    // This way, we can inversely correlate height difference with speed
+    ASSERT(m_apex - m_midpoint.y > std::numeric_limits<f32>::epsilon());
+    f32 maxVel = static_cast<f32>(static_cast<s16>(params.setting(0)));
+    m_maxVelSq = maxVel * maxVel;
+    m_accel = m_maxVelSq / (m_apex - m_midpoint.y);
+}
+
+/// @addr{0x806CEB24}
+ObjectHeyho::~ObjectHeyho() = default;
+
+/// @addr{0x806CEB90}
+void ObjectHeyho::init() {
+    ASSERT(m_railInterpolator);
+    m_railInterpolator->init(0.0f, m_railInterpolator->pointCount() / 2);
+    m_pos = m_railInterpolator->curPos();
+    m_flags.setBit(eFlags::Position);
+    m_currentVel = EGG::Mathf::sqrt(
+            m_maxVelSq - m_accel * (m_railInterpolator->curPos().y - m_midpoint.y));
+
+    m_transformOffset = m_railInterpolator->curTangentDir() * m_currentVel;
+    m_floorCollision = false;
+    m_up = EGG::Vector3f::ey;
+    m_forward = EGG::Vector3f::ez;
+    m_freeFall = false;
+    m_spinFrame = 0;
+
+    changeAnimation(Animation::Move);
+}
+
+/// @addr{0x806CEDF8}
+void ObjectHeyho::calc() {
+    calcStateTransition();
+    calcMotion();
+    StateManager::calc();
+    calcInterp();
+}
+
+/// @addr{0x806D013C}
+void ObjectHeyho::loadAnims() {
+    std::array<const char *, 4> names = {{
+            "body_color",
+            "move",
+            "jump",
+            "jump_ed",
+    }};
+
+    std::array<Render::AnmType, 4> types = {{
+            Render::AnmType::Pat,
+            Render::AnmType::Chr,
+            Render::AnmType::Chr,
+            Render::AnmType::Chr,
+    }};
+
+    linkAnims(names, types);
+}
+
+/// @addr{0x806D01D4}
+void ObjectHeyho::calcCollisionTransform() {
+    auto *objCol = collision();
+    if (!objCol) {
+        return;
+    }
+
+    EGG::Matrix34f tm;
+    tm.makeT(EGG::Vector3f(0.0f, 100.0f, 0.0f));
+    calcTransform();
+    EGG::Matrix34f m = m_transform.multiplyTo(tm);
+    objCol->transform(m, m_scale, m_transformOffset);
+}
+
+/// @addr{0x806CF4D0}
+void ObjectHeyho::calcMove() {
+    m_forward = m_railInterpolator->nextPoint().pos - m_railInterpolator->curPoint().pos;
+    m_floorCollision = false;
+
+    CollisionInfo info;
+
+    if (CollisionDirector::Instance()->checkSphereFull(100.0f, m_pos + COLLISION_OFFSET,
```

Comment:
Add a `static constexpr f32 RADIUS = 100.0f;` to the header. Then you can also use this in `calcJump()`